design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1072 @@
+string keys).
+
+For example, given a statement such as: 
+
+```
+CREATE STREAM USERS (NAME STRING) WITH (...);
+ ``` 
+ 
+The logical value schema has only a single field. KSQL should be able to read 
+data in Kafka where the value is a anonymous top-level JSON or Avro string, rather 
+than only a _record_ containing a `NAME` field.
+
+Given a statement such as:
+
+```
+CREATE STREAM FOO AS SELECT ID FROM BAR;
+ ``` 
+ 
+The _logical_ schema is again only a single field, lets say of type `INT`. 
+Users should be able to control if this is serialized as a named `ID` field within
+_record_, or as an anonymous JSON number or Avro int. 
+ 
+
+## What is in scope
+
+* JSON (de)serialization of schemas containing a single value field that 
+  is an `ARRAY`, `MAP`, `STRUCT` or primitive type.
+  
+* Avro (de)serialization of schemas containing a single value field that
+  is an `ARRAY`, `MAP`, `STRUCT` or primitive type.  
+  
+* (De)serialization of _record_ and _non-record_ keys, but only in the context on
+  ensuring a design that can be extended to include such keys.
+  
+* Any configuration and/or syntax used to control the (de)serialization of single fields.
+
+## What is not in scope
+
+* Extensive details of how KSQL will be extended to support all _record_ and 
+  _non-record_ key types. This is future work. 
+  
+* Other current or future formats. Though the design needs to be mindful to 
+  ensure functionality is compatible with future formats that KSQL may support.
+
+## Value/Return
+
+This functionality will allow KSQL to work with more _value_ schemas, opening 
+up KSQL to more use-cases.
+
+This work is also a precondition for structured key support, which will allow
+KSQL to work with more _key_ schemas, opening up KSQL to even more use-cases. 
+
+## Public APIS
+
+### New compatibility breaking config:
+
+* `ksql.persistence.wrap.single.values`: which provides a default for how 
+  VALUE schemas with a single field should be deserialized and serialized:
+  `true` indicating values should be persisted within a _record_; `false`
+  indicating values should be persisted as anonymous values.
+  
+  With a legacy default of `true` to ensure backwards compatibility of old
+  queries, and a new default of `false`, meaning new queries will default
+  to writing single value fields as anonymous values.
+  
+* `ksql.persistence.wrap.single.keys`: which provides a default for how 
+  KEY schemas with a single field should be deserialized and serialized:
+  `true` indicating keys should be persisted within a _record_; `false`
+  indicating values should be persisted as anonymous values.
+  
+  With a legacy default of `true` to ensure backwards compatibility of old
+  queries with string keys, and a new default of `false`, meaning new 
+  queries will default to writing single value fields as anonymous values.
+  
+  Note: to be added as part of the future structured keys work.
```

Comment:
_what_ is to be added as part of the future work? it's unclear the scope to which this note applies

Reply:
It's scoped to the bullet point its indented with. I think this will be clear if you read the rendered MD, rather than read the source.